prover/src/matrix/row_matrix.rs

```diff
@@ -0,0 +1,200 @@
+// Copyright (c) Facebook, Inc. and its affiliates.
+//
+// This source code is licensed under the MIT license found in the
+// LICENSE file in the root directory of this source tree.
+
+use super::segments::Segment;
+use crate::{matrix::ARR_SIZE, Matrix};
+use math::{fft, FieldElement, StarkField};
+use utils::collections::Vec;
+use utils::{flatten_vector_elements, uninit_vector};
+
+// ROWMAJOR MATRIX
+// ================================================================================================
+
+/// A row-major matrix of field elements. The matrix is represented as a single vector of field
+/// elements, where the first `row_len` elements represent the first row of the matrix, the next
+/// `row_len` elements represent the second row, and so on.
+///
+/// # Note
+/// - The number of rows in the matrix is always a multiple of the ARR_SIZE.
+/// - The number of columns in the matrix is always a multiple of the ARR_SIZE.
+#[derive(Clone, Debug)]
+pub struct RowMatrix<E: FieldElement> {
+    data: Vec<E>,
+    row_len: usize,
+}
+
+impl<E> RowMatrix<E>
+where
+    E: FieldElement,
+{
+    // CONSTRUCTORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Creates a new row-major matrix from the specified data and row length. The data must be
+    /// arranged in row-major order, i.e. the first `row_len` elements of the data represent the first
+    /// row of the matrix, the next `row_len` elements represent the second row, and so on.
+    ///
+    /// # Panics
+    /// - if the number of elements in the data is not a multiple of the specified row length;
+    /// - if the specified row length is not a multiple of the ARR_SIZE;
+    /// - if the specified data is empty.
+    pub fn new(data: Vec<E>, row_len: usize) -> Self {
+        assert!(data.len() % row_len == 0);
+        assert!(row_len % ARR_SIZE == 0);
+        assert!(!data.is_empty());
+
+        Self { data, row_len }
+    }
+
+    /// Converts a column-major matrix of polynomials `Matrix<E>` into a RowMatrix evaluated at
+    /// a shifted domain offset.
+    pub fn from_polys(polys: &Matrix<E>, blowup_factor: usize) -> Self {
```

Comment:
Although I get the idea, I find the naming / description slightly off point / confusing. But I'm not sure what would be really best, so maybe we can discard my comment

Reply:
I was thinking of using `from_col_matrix` but find the current one better.

Reply:
This basically boils down to a matrix transposition followed by an LDE extension, right? Maybe we could try something reflecting this, like `pub fn transpose_and_extend()`?

Reply:
done! I've updated it here. Please see [this](https://github.com/facebook/winterfell/blob/19be088ab4849cc34de3db0a1821db5364d30164/prover/src/matrix/row_matrix.rs#L53)

Reply:
Thanks!